File: atomasm.c
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#define BUF_SIZE 256

// Errors
#define UNKNOWN_MNEMONIC_ERROR 1
#define FILE_NOT_FOUND_ERROR 2
#define NOT_ENOUGH_ARGUMENTS_ERROR 3
#define OUTPUT_ERROR 4

// Opcodes
#define HALT 0
#define PUSH 1
#define POP 2
#define ADD 3
#define SUB 4
#define MUL 5
#define DIV 6
#define JMP 7
#define JIG 8
#define JIE 9
#define JIS 10
#define JIZ 11
#define JNE 12
#define PUTN 13
#define PUTC 14
#define JNZ 15
#define DUP 16
#define STORE 17
#define LOAD 18

int has_prefix(char *pre, char *str) {
    return strncmp(pre, str, strlen(pre)) == 0;
}

int line_to_bytecode(char *line, uint8_t *byte_buf, size_t *top_idx) {
    if (strlen(line) == 0) {
        return 0;
    }

    if (has_prefix("HALT", line)) {
        byte_buf[*top_idx] = HALT;
        *top_idx++;
        return 0;
    }

    return UNKNOWN_MNEMONIC_ERROR;
}

int main(int argc, char **argv) {
    if (argc < 3) {
        printf("Usage: atomasm <input file> <output file>\n");
        return NOT_ENOUGH_ARGUMENTS_ERROR;
    }

    char *file_name = argv[1];
    char *output_path = argv[2];
    char line_buf[BUF_SIZE] = {0};
    uint8_t byte_buf[BUF_SIZE] = {0};
    size_t top_idx = 0;

    FILE *file = fopen(file_name, "r");
    if (file == NULL) {
        printf("File '%s' not found\n", file_name);
        return FILE_NOT_FOUND_ERROR;
    }

    int idx = 1;
    while (fgets(line_buf, BUF_SIZE, file)) {
        int code = line_to_bytecode(line_buf, byte_buf, &top_idx);
        if (code) {
            printf("ERROR WHILE ASSEMBLING BYTECODE AT LINE: %d\n", idx);
            printf("REASON: ");
            switch (code) {
                case UNKNOWN_MNEMONIC_ERROR:
                    printf("UNKNOWN MNEMONIC\n");
                    break;

                default:
                    printf("UNKNOWN ERROR\n");
                    break;
            }

            return code;
        }

        idx++;
    }

    fclose(file);

    FILE *output_file = fopen(output_path, "w");
    if (output_file == NULL) {
        printf("Error while writing to output file\n");
        return OUTPUT_ERROR;
    }

    fwrite(byte_buf, sizeof(uint8_t), BUF_SIZE, output_file);
    fclose(output_file);

    return 0;
}

File: atomasm.nim
import std/cmdline
import std/strutils
import std/tables

proc gatherLabels(tokens: seq[string]): TableRef[string, int] =
  var labels = newTable[string, int]()
  var idx = 0

  for token in tokens:
    if token.toLower() == "label":
      labels[tokens[idx + 1]] = idx

    idx += 1

  return labels

proc tokensToByteCode(tokens: seq[string], labels: TableRef): seq[byte] =
  var idx = 0
  var ignorable: seq[int] = @[]
  var bytes: seq[byte] = @[]

  for token in tokens:
    if token.toLower().startsWith("label"):
      idx += 1
      continue

    if ignorable.contains(idx):
      idx += 1
      continue

    case token.toLower() 
      of "halt":
        bytes.add(0x0)

      of "push":
        bytes.add(0x1)
        let next = tokens[idx + 1]
        let val = parseInt(next)
        if val < 0 or val > 255:
          raise newException(Exception, "Value of immediate does not fit in a byte")
        else:
          bytes.add(cast[byte](val))

      of "putn":
        bytes.add(0xd)
          
      of "putc":
        bytes.add(0xe)

      of "pop":
        bytes.add(0x2)
            
      of "add":
        bytes.add(0x3)
                
      of "sub":
        bytes.add(0x4)

      of "mul":
        bytes.add(0x5)

      of "div":
        bytes.add(0x6)

      of "jmp":
        bytes.add(0x7)
        let next = tokens[idx + 1]
        bytes.add(cast[byte](labels[next]))

      of "jig":
        bytes.add(0x8)
        let next = tokens[idx + 1]
        bytes.add(cast[byte](labels[next]))

      of "jis":
        bytes.add(0xa)
        let next = tokens[idx + 1]
        bytes.add(cast[byte](labels[next]))

      of "jie":
        bytes.add(0x9)
        let next = tokens[idx + 1]
        bytes.add(cast[byte](labels[next]))

      of "jiz":
        bytes.add(0xb)
        let next = tokens[idx + 1]
        bytes.add(cast[byte](labels[next]))

      of "jne":
        bytes.add(0xc)
        let next = tokens[idx + 1]
        bytes.add(cast[byte](labels[next]))

      of "dup":
        bytes.add(0x10)

      of "store":
        bytes.add(0x11)
        let next = tokens[idx + 1]
        let val = parseInt(next)
        if val < 0 or val > 255:
          raise newException(Exception, "Value of immediate does not fit in a byte")
        else:
          bytes.add(cast[byte](val))

      of "load":
        bytes.add(0x12)
        let next = tokens[idx + 1]
        let val = parseInt(next)
        if val < 0 or val > 255:
          raise newException(Exception, "Value of immediate does not fit in a byte")
        else:
          bytes.add(cast[byte](val))

    idx += 1

  return bytes


proc main(args: seq[string]): int =
  let args = commandLineParams()
  if args.len < 2:
    echo "Usage: atomasm <input file> <output file>"
    return 1

  let input_path = args[0]
  let output_path = args[1]
  var content: string

  try:
    content = readFile(input_path)
  except IOError:
    echo "Could not open input file"
    return 1

  let tokens = strutils.splitWhitespace(content, -1)
  let labels = gatherLabels(tokens)

  try:
    let bytes = tokensToByteCode(tokens, labels)
    writeFile(output_path, bytes)
  except Exception as e:
    echo "Error while generating bytecode/writing to file: " .. e.msg
    return 1

  return 0

quit(main(commandLineParams()))
